vsintegration/src/FSharp.Editor/Common/OpenDeclarationsGetter.fs

```diff
@@ -0,0 +1,476 @@
+﻿namespace Microsoft.VisualStudio.FSharp.Editor
+
+open System
+open System.Collections.Generic
+open Microsoft.FSharp.Compiler
+open Microsoft.FSharp.Compiler.Ast
+open Microsoft.FSharp.Compiler.SourceCodeServices
+open Microsoft.FSharp.Compiler.Range
+
+type RawOpenDeclaration =
+    { Idents: Idents
+      Parent: Idents option }
+
+type OpenDeclWithAutoOpens =
+    { Declarations: Idents list
+      Parent: Idents option
+      IsUsed: bool }
+
+[<NoComparison>]
+type internal OpenDeclaration =
+    { Declarations: OpenDeclWithAutoOpens list
+      DeclarationRange: Range.range
+      ScopeRange: Range.range
+      IsUsed: bool }
+
+[<CompilationRepresentation (CompilationRepresentationFlags.ModuleSuffix)>]
+module OpenDeclWithAutoOpens =
+    let updateBySymbolPrefix (symbolPrefix: Idents) (decl: OpenDeclWithAutoOpens) =
+        let matched = decl.Declarations |> List.exists (Array.areEqual symbolPrefix)
+        //if not decl.IsUsed && matched then debug "OpenDeclarationWithAutoOpens %A is used by %A" decl symbolPrefix
+        matched, { decl with IsUsed = decl.IsUsed || matched }
+
+[<CompilationRepresentation (CompilationRepresentationFlags.ModuleSuffix)>]
+module internal OpenDeclaration =
+    let updateBySymbolPrefix (symbolPrefix: Idents) (decl: OpenDeclaration) =
+        let decls =
+            decl.Declarations 
+            |> List.map (OpenDeclWithAutoOpens.updateBySymbolPrefix symbolPrefix)
+        let matched = decls |> List.exists fst
+        let isUsed = decls |> List.exists (fun (_, decl) -> decl.IsUsed)
+        //if not decl.IsUsed && isUsed then debug "OpenDeclaration %A is used by %A" decl symbolPrefix
+        matched, { decl with Declarations = decls |> List.map snd; IsUsed = isUsed }
+
+module internal OpenDeclarationGetter =
+    open System 
+    open System.Diagnostics
+
+    let private getAutoOpenModules entities =
+        List.foldBack( fun (e: RawEntity) acc -> 
+             match e.Kind LookupType.Precise with
+             |  EntityKind.Module { IsAutoOpen = true } -> e.CleanedIdents :: acc
+             | _ -> acc) entities []
+
+
+    let private getActivePatterns (entities: RawEntity list) =
+        List.foldBack( fun (e: RawEntity) acc -> 
```

Comment:
This is a ``List.filter`` or ``List.choose``.  Please don't use ``fold`` operations when a better alternative is available.